src/connection/adapter/base_http_adapter.ts

```diff
@@ -107,20 +105,32 @@ export abstract class BaseHttpAdapter implements Connection {
 
   protected abstract createClientRequest(
     url: URL,
-    params: RequestParams
+    params: RequestParams & { abort_controller: AbortController }
   ): Http.ClientRequest
 
   protected async request(params: RequestParams): Promise<Stream.Readable> {
     return new Promise((resolve, reject) => {
       const start = Date.now()
 
-      const request = this.createClientRequest(params.url, params)
-      request.once('socket', (socket) => {
-        socket.setTimeout(this.config.request_timeout)
+      const abortController = params?.abort_controller || new AbortController()
+      let isTimedOut = false
+      const timeout = setTimeout(() => {
+        isTimedOut = true
+        abortController.abort('Request timed out')
```

Comment:
How does `abortController` behaves if external code calls `abort` on the already aborted controller?

Reply:
Like, timeout and then abort?

Reply:
yes, for example. the chance is small though